### PR TITLE
[AAG-25] configure project dependencies and basic setup

### PR DESCRIPTION
**Pull request by dineshnukemars**
%Please include a summary of the change or/and which issue is fixed + relevant motivation and context.%

[AAG-25](https://distillery.atlassian.net/browse/AAG-25)

**Type of change:**
- Other: basic configurations and Gradle dependencies

**How Has This Been Tested?**
- built and launched a project

**Checklist:**
- [x] I have performed a self-review of my own code
- [x] I documented all new code and updated the external docs if necessary
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: dependencies.gradle [hunk 1]
```diff
@@ -15,21 +15,40 @@ project.ext {
...
+                    kotlinCoroutinesAndroid: "org.jetbrains.kotlinx:kotlinx-coroutines-android:$versions.platform.coroutines",
+                    gradle                 : "com.android.tools.build:gradle:4.0.0",
             ],
             android : [
                     core                     : "androidx.core:core-ktx:1.2.0",
                     supportAppCompat         : "androidx.appcompat:appcompat:1.1.0",
                     supportCardView          : "androidx.cardview:cardview:1.0.0",
                     supportConstraintLayout  : "androidx.constraintlayout:constraintlayout:1.1.3",
-                    supportLifecycleViewModel: "androidx.lifecycle:lifecycle-viewmodel-ktx:$versions.android.lifecycleVersion",
+                    supportLifecycleViewModel: "androidx.lifecycle:lifecycle-viewmodel-ktx:2.2.0",
```
Comment (cnavarrete101) [inline dependencies.gradle:29]: why didn't modify or create a variable for this one?

File: dependencies.gradle [hunk 2]
```diff
@@ -15,21 +15,40 @@ project.ext {
...
+            koin    : [
+                    core      : "org.koin:koin-core:$versions.android.koin",
+                    coreExt   : "org.koin:koin-core-ext:$versions.android.koin",
+                    test      : "org.koin:koin-test:$versions.android.koin",
+                    scope     : "org.koin:koin-androidx-scope:$versions.android.koin",
+                    viewModel : "org.koin:koin-androidx-viewmodel:$versions.android.koin",
+                    androidExt: "org.koin:koin-androidx-ext:$versions.android.koin"
+            ],
+            gson    : [
+                    core: "com.google.code.gson:gson:2.8.6"
```
Comment (anaumchik) [inline dependencies.gradle:51]: We don't need gson. Because we won't have network calls.

File: build.gradle [hunk 3]
```diff
@@ -14,6 +14,7 @@ buildscript {
 
         // for detekt plugin
         classpath "io.gitlab.arturbosch.detekt:detekt-gradle-plugin:1.7.4"
+        classpath 'com.android.tools.build:gradle:4.0.0'
```
Comment (anaumchik) [inline build.gradle:17]: why we required to add gradle here?

File: app/src/main/java/com/distillery/android/blueprints/MainApplication.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+import android.app.Application
+import com.distillery.android.domain.Dependencies
+
+class MainApplication : Application() {
+
+    override fun onCreate() {
+        super.onCreate()
+
+        val diImplementations = listOf<Dependencies>()// add all the implementation EX: mviDI, MvpDI
```
Comment (anaumchik) [inline app/src/main/java/com/distillery/android/blueprints/MainApplication.kt:11]: let's name this val just `dependencies`

File: app/src/main/res/layout/activity_main.xml [hunk 5]
```diff
@@ -7,13 +7,50 @@
     android:layout_height="match_parent"
     tools:context=".MainActivity"
     >
-    <TextView
-        android:layout_width="wrap_content"
+    <Button
+        android:id="@+id/view_MVI_Entry"
```
Comment (anaumchik) [inline app/src/main/res/layout/activity_main.xml:11]: with Kotlin synthetic it's better name id with snakeCase `btnEntryPointMVI` or `btnEntryMVI`

the same for MVP and MVVM in this file

File: app/src/main/res/layout/activity_main.xml [hunk 6]
```diff
@@ -7,13 +7,50 @@
...
+        android:layout_width="200dp"
         android:layout_height="wrap_content"
-        android:text="Hello World!"
+        android:layout_marginTop="16dp"
+        android:layout_marginEnd="16dp"
+        android:layout_marginBottom="16dp"
+        android:layout_marginStart="16dp"
+        android:text="MVI Entry"
```
Comment (anaumchik) [inline app/src/main/res/layout/activity_main.xml:18]: text to resources

the same in this file

File: app/src/main/res/layout/activity_main.xml [hunk 7]
```diff
@@ -7,13 +7,50 @@
...
+        android:id="@+id/textView3"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="20dp"
+        android:layout_marginStart="24dp"
+        android:text="This is just temporary UI"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/view_MVP_Entry"
         />
 </androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (anaumchik) [inline app/src/main/res/layout/activity_main.xml:56]: @distillery-alexander-ageychenko @0neel do we need such king of entry point screen? If yes could we create specific design for it?

File: mvi_example/src/main/java/com/distillery/android/blueprints/mvi/MviActivity.kt [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package com.distillery.android.blueprints.mvi
+
+import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+
+class MviActivity : AppCompatActivity() {
```
Comment (anaumchik) [inline mvi_example/src/main/java/com/distillery/android/blueprints/mvi/MviActivity.kt:6]: It any file in `mvi` package created as a template, right?

File: app/src/main/java/com/distillery/android/blueprints/MainApplication.kt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+        diImplementations.forEach {
+            callDI(it)
+        }
+    }
+    private fun callDI(dependencies: Dependencies) {
```
Comment (distillery-alexander-ageychenko) [inline app/src/main/java/com/distillery/android/blueprints/MainApplication.kt:16]: Better to avoid using abbreviation is the methods names, also we are using `camelCase` and it is applied to abbreviations too, so it should be at least `callDi`.
What does it do? As I see it doesn't call anything, but rather loads dependencies, so naming can be improved.

File: app/src/main/AndroidManifest.xml [hunk 10]
```diff
@@ -1,17 +1,16 @@
 <?xml version="1.0" encoding="utf-8"?>
-<manifest
-    xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.distillery.android.blueprints"
-    >
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.distillery.android.blueprints">
```
Comment (distillery-alexander-ageychenko) [inline app/src/main/AndroidManifest.xml:3]: Please revert these changes, thay are unrelated and previous formatting was there for a reason. It makes less diff when changing last line.

File: app/src/main/AndroidManifest.xml [hunk 11]
```diff
@@ -1,17 +1,16 @@
 <?xml version="1.0" encoding="utf-8"?>
-<manifest
-    xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.distillery.android.blueprints"
-    >
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (distillery-alexander-ageychenko) [inline app/src/main/AndroidManifest.xml:2]: Please revert unrelated changes. Also previous formatting was there for a reason, it helps to reduce diff lines when last line is being changed

File: app/src/main/java/com/distillery/android/blueprints/MainActivity.kt [hunk 12]
```diff
@@ -1,12 +1,22 @@
...
         setContentView(R.layout.activity_main)
+
+        view_MVI_Entry.setOnClickListener {
+        }
+
+        view_MVVM_Entry.setOnClickListener {
+        }
+
+        view_MVP_Entry.setOnClickListener {
+        }
```
Comment (distillery-alexander-ageychenko) [inline app/src/main/java/com/distillery/android/blueprints/MainActivity.kt:20]: Do we need these entries in this PR? They should be added at a time of usage as I think.

File: app/src/main/res/layout/activity_main.xml [hunk 13]
```diff
@@ -7,13 +7,50 @@
     android:layout_height="match_parent"
     tools:context=".MainActivity"
     >
-    <TextView
-        android:layout_width="wrap_content"
+    <Button
+        android:id="@+id/view_MVI_Entry"
+        android:layout_width="200dp"
```
Comment (distillery-alexander-ageychenko) [inline app/src/main/res/layout/activity_main.xml:12]: It's better to avoid such sizes in layouts. Making it match parent with margins will look much better on different devices

File: app/src/main/res/layout/activity_main.xml [hunk 14]
```diff
@@ -7,13 +7,50 @@
...
         android:layout_height="wrap_content"
-        android:text="Hello World!"
+        android:layout_marginTop="16dp"
```
Comment (distillery-alexander-ageychenko) [inline app/src/main/res/layout/activity_main.xml:14]: Better to use resources in such case, as same margins could be applied in lots of places.

File: app/src/main/res/layout/activity_main.xml [hunk 15]
```diff
@@ -7,13 +7,50 @@
...
+        android:layout_marginEnd="16dp"
+        android:layout_marginBottom="16dp"
+        android:layout_marginStart="16dp"
+        android:text="MVI Entry"
         app:layout_constraintBottom_toBottomOf="parent"
-        app:layout_constraintLeft_toLeftOf="parent"
-        app:layout_constraintRight_toRightOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintHorizontal_bias="0.035"
```
Comment (distillery-alexander-ageychenko) [inline app/src/main/res/layout/activity_main.xml:21]: Could you explain such variable? Does it have any meaning or what it's change will affect?

File: app/src/main/res/layout/activity_main.xml (hunk 7 above)
Comment (distillery-alexander-ageychenko) [reply app/src/main/res/layout/activity_main.xml:56]: @anaumchik makes sense, but for now we can just change base activity in manifest instead to launch required architecture.

File: dependencies.gradle [hunk 16]
```diff
@@ -15,21 +15,40 @@ project.ext {
     ]
     libraries = [
             platform: [
-                    kotlinGradlePlugin: "org.jetbrains.kotlin:kotlin-gradle-plugin:$versions.platform.kotlin",
-                    kotlinStdLib      : "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$versions.platform.kotlin",
-                    kotlinCoroutines  : "org.jetbrains.kotlinx:kotlinx-coroutines-core:$versions.platform.coroutines",
-                    gradle            : "com.android.tools.build:gradle:3.5.3",
+                    kotlinGradlePlugin     : "org.jetbrains.kotlin:kotlin-gradle-plugin:$versions.platform.kotlin",
```
Comment (distillery-alexander-ageychenko) [inline dependencies.gradle:18]: new dependencies are not connected with new screen, please move it to another PR so one PR have 1 goal.

File: domain/src/main/java/com/distillery/android/domain/Dependencies.kt [hunk 17]
```diff
@@ -0,0 +1,8 @@
+package com.distillery.android.domain
+
+import android.content.Context
+
+interface Dependencies {
+
+    fun loadDependencies(androidContext: Context)
```
Comment (distillery-alexander-ageychenko) [inline domain/src/main/java/com/distillery/android/domain/Dependencies.kt:7]: There's no need to specify that it is an android context, we don't have any other. In case it should be instance of activity - then better to narrow down the argument type to activity.
```suggestion
    fun loadDependencies(content: Context)
```

Comment (distillery-alexander-ageychenko) [inline domain/src/main/java/com/distillery/android/domain/Dependencies.kt:7]: Major: Note that interface is in domain layer but uses platform specific class which is violates dependency inversion rule. Context should be used in implementation on a data layer, probably passed in a constructor.

File: domain/src/main/java/com/distillery/android/domain/FakeToDoRepository.kt [hunk 18]
```diff
@@ -122,4 +122,11 @@ class FakeToDoRepository : ToDoRepository {
         }
         todosChannel.close(IllegalArgumentException("You died"))
     }
+
+    override suspend fun deleteToDo(uniqueId: Long) {
```
Comment (distillery-alexander-ageychenko) [inline domain/src/main/java/com/distillery/android/domain/FakeToDoRepository.kt:126]: adding a new method is not connected with libraries updates and adding of a screen elements, move it to another PR so one PR have one goal.

File: domain/src/main/java/com/distillery/android/domain/FakeToDoRepository.kt [hunk 19]
```diff
@@ -24,18 +24,18 @@ private val idGenerator = AtomicLong(0)
...
-class FakeToDoRepository : ToDoRepository {
+class FakeToDoRepository(private val scope: CoroutineScope) : ToDoRepository {
 
     private val toDos = ConcurrentHashMap<Long, ToDoModel>()
     private val todosChannel = Channel<List<ToDoModel>>()
 
     // generate infinite amount of elements, but throw exception after 10th element
     init {
-        GlobalScope.launch {
+        scope.launch {
```
Comment (distillery-alexander-ageychenko) [inline domain/src/main/java/com/distillery/android/domain/FakeToDoRepository.kt:38]: replacing scope is yet another goal for this PR, which makes it impossible to revert in case of any issues

File: mvi_example/src/main/res/values/strings.xml [hunk 20]
```diff
@@ -0,0 +1,10 @@
+<resources>
+    <string name="app_name">Distillery Android Architecture Blueprints</string>
+    <string name="title">Title</string>
```
Comment (distillery-alexander-ageychenko) [inline mvi_example/src/main/res/values/strings.xml:3]: common UI resources were added in another PR already, please sync with @sasfmlzr

File: dependencies.gradle (hunk 1 above)
Comment (dineshnukemars) [reply dependencies.gradle:29]: I did it earlier, I think the changes got lost while rebasing

File: mvi_example/src/main/java/com/distillery/android/blueprints/mvi/MviActivity.kt (hunk 8 above)
Comment (dineshnukemars) [reply mvi_example/src/main/java/com/distillery/android/blueprints/mvi/MviActivity.kt:6]: I don't understand, could you please explain?

File: app/src/main/res/layout/activity_main.xml (hunk 15 above)
Comment (dineshnukemars) [reply app/src/main/res/layout/activity_main.xml:21]: couldn't understand what you are asking. could you explain?

File: dependencies.gradle (hunk 16 above)
Comment (dineshnukemars) [reply dependencies.gradle:18]: the pr goal is to update the dependencies or in other words to set up basic environment for developers. As I discussed with my team, common dependencies should be separate PR.

File: domain/src/main/java/com/distillery/android/domain/Dependencies.kt (hunk 17 above)
Comment (dineshnukemars) [reply domain/src/main/java/com/distillery/android/domain/Dependencies.kt:7]: resolved

File: domain/src/main/java/com/distillery/android/domain/FakeToDoRepository.kt (hunk 18 above)
Comment (dineshnukemars) [reply domain/src/main/java/com/distillery/android/domain/FakeToDoRepository.kt:126]: I didn't add anything new, this probably a mistake while reformatting. will fix it

File: domain/src/main/java/com/distillery/android/domain/FakeToDoRepository.kt (hunk 19 above)
Comment (dineshnukemars) [reply domain/src/main/java/com/distillery/android/domain/FakeToDoRepository.kt:38]: not my changes, this is a difference between master and development. I will change the target branch to a development

File: mvi_example/src/main/res/values/strings.xml (hunk 20 above)
Comment (dineshnukemars) [reply mvi_example/src/main/res/values/strings.xml:3]: not my changes, this is a difference between master and development. I will change the target branch to a development

**Comment by dineshnukemars** [comment]
most of the change list is because of rebasing error. I will raise new PR, hence closing this. thx a lot for the review.
